published/external/xdpapi.h

```diff
@@ -156,7 +228,269 @@ XdpUnloadApi(
+    )
+{
+    XDP_API_CLIENT *_Client = NULL;
+    NTSTATUS _Status;
+    NPI_CLIENT_CHARACTERISTICS *_NpiCharacteristics;
+    NPI_REGISTRATION_INSTANCE *_NpiInstance;
+    BOOLEAN _ResourceInitialized = FALSE;
+
+    if (_XdpApiVersion != XDP_API_VERSION_1) {
+        _Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    _Client =
+        (XDP_API_CLIENT *)ExAllocatePoolZero(
+            NonPagedPoolNx, sizeof(*_Client), POOLTAG_NMR_CLIENT);
+    if (_Client == NULL) {
+        _Status = STATUS_NO_MEMORY;
+        goto Exit;
+    }
+
+    //
+    // Use resources instead of push locks for downlevel compatibility.
+    //
+    _Status = ExInitializeResourceLite(&_Client->Resource);
+    if (!NT_SUCCESS(_Status)) {
+        goto Exit;
+    }
+    _ResourceInitialized = TRUE;
+
+    _Client->ModuleId.Length = sizeof(_Client->ModuleId);
+    _Client->ModuleId.Type = MIT_GUID;
+    XdpGuidCreate(&_Client->ModuleId.Guid);
+
+    _Client->Context = _ClientContext;
+    _Client->Attach = _ClientAttach;
+    _Client->Detach = _ClientDetach;
+    _Client->XdpApiClientDispatch = _XdpApiClientDispatch;
+
+    _NpiCharacteristics = &_Client->NpiClientCharacteristics;
+    _NpiCharacteristics->Length = sizeof(*_NpiCharacteristics);
+    _NpiCharacteristics->ClientAttachProvider = XdpNmrClientAttachProvider;
+    _NpiCharacteristics->ClientDetachProvider = XdpNmrClientDetachProvider;
+
+    _NpiInstance = &_NpiCharacteristics->ClientRegistrationInstance;
+    _NpiInstance->Size = sizeof(*_NpiInstance);
+    _NpiInstance->Version = 0;
+    _NpiInstance->NpiId = &NPI_XDPAPI_INTERFACE_ID;
+    _NpiInstance->ModuleId = &_Client->ModuleId;
+    _NpiInstance->Number = XDP_API_VERSION_1;
+
+    _Status =
+        NmrRegisterClient(
+            &_Client->NpiClientCharacteristics, _Client, &_Client->NmrClientHandle);
+    if (!NT_SUCCESS(_Status)) {
+        goto Exit;
+    }
+
+    *_XdpLoadApiContext = (XDP_LOAD_API_CONTEXT)_Client;
+    _Status = STATUS_SUCCESS;
+
+Exit:
+
+    if (!NT_SUCCESS(_Status)) {
+        if (_Client != NULL) {
+            XdpCleanupClientRegistration(_Client, _ResourceInitialized);
+            ExFreePoolWithTag(_Client, POOLTAG_NMR_CLIENT);
+        }
+    }
+
+    return _Status;
+}
+
+inline
+VOID
+XdpUnloadApi(
+    _In_ XDP_LOAD_API_CONTEXT _XdpLoadApiContext
+    )
+{
+    XDP_API_CLIENT *_Client = (XDP_API_CLIENT *)_XdpLoadApiContext;
+
+    XdpCleanupClientRegistration(_Client, TRUE);
+    ExFreePoolWithTag(_Client, POOLTAG_NMR_CLIENT);
+}
+
+inline
+NTSTATUS
+XdpOpenApi(
+    _In_ XDP_LOAD_API_CONTEXT _XdpLoadApiContext,
+    _Out_ const XDP_API_PROVIDER_DISPATCH **_XdpApiProviderDispatch,
+    _Out_ const XDP_API_PROVIDER_BINDING_CONTEXT **_XdpApiProviderContext
+    )
+{
+    XDP_API_CLIENT *_Client = (XDP_API_CLIENT *)_XdpLoadApiContext;
+    NTSTATUS _Status;
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&_Client->Resource);
+
+    if (_Client->BindingHandle == NULL) {
```

Comment:
same comment as XdpNmrClientDetachProvider - how do we prevent NMR unloading xdp from under the client?

Reply:
@nigriMSFT ?

Reply:
Right, returning success inline in the Detach routine was a prototype shortcut. We need some sort of notification mechanism to let the client know that XDP is unloading and that the client should clean up. The Detach routine should initiate this notification and should return PENDING. When the client is done cleaning up anything it needs to (waiting for any outstanding API calls to complete, closing XDPAPI object handles, etc.), it should complete the Detach operation.

Reply:
split PR
https://github.com/microsoft/xdp-for-windows/pull/719